### ISSUE TITLE
bndlib requires OSGI 6.0.0, but fails to compile due to missing abstract methods added in 6.0.0

### ISSUE THREAD

**Issue by wltjr** [bndtools/bnd#1322]
I am trying to compile just bndlib, tag 3.1.0, stand alone. Compiling java code under bnd/biz.aQute.bndlib/src only nothing else. I tried from master and seems to have same issue, not sure if it is address in another tag.

osgi 6.0.0 is required to compile, but bndlib is missing some new methods that were added in org.osgi.service.repository.Repository class, that are not implemented in bndlib.

public ExpressionCombiner getExpressionCombiner()
public RequirementBuilder newRequirementBuilder(String namespace)

I am getting the following error. There are more errors, it just stops at the first missing abstract method, there is a 2nd missing one, getExpressionCombiner().

src/aQute/bnd/osgi/repository/ResourcesRepository.java:15: error: ResourcesRepository is not abstract and does not override abstract method newRequirementBuilder(String) in Repository
public class ResourcesRepository implements org.osgi.service.repository.Repository {
       ^
src/aQute/bnd/osgi/repository/AugmentRepository.java:23: error: AugmentRepository is not abstract and does not override abstract method newRequirementBuilder(String) in Repository
public class AugmentRepository implements Repository {
       ^
src/aQute/bnd/osgi/repository/AggregateRepository.java:15: error: AggregateRepository is not abstract and does not override abstract method newRequirementBuilder(String) in Repository
public class AggregateRepository implements Repository {
       ^
3 errors
1 warning

It does not seem to be implemented in the latest sources. But the sources also require osgi 6.0.0 as I get many other errors about missing annotation.* classes when I use older versions of osgi. Those classes do not exist in <=5.0.0.

Are there any plans to implement these new methods? I was looking at jbosgi-repository, which has code that the new methods call. Not sure if that can be used, or needs to be re-coded for bndlib.

These two classes
https://github.com/jbosgi/jbosgi-repository/tree/master/core/src/main/java/org/jboss/osgi/repository/internal

The new methods are just wrappers around those two
https://github.com/jbosgi/jbosgi-repository/blob/master/core/src/main/java/org/jboss/osgi/repository/spi/AbstractRepository.java

    @Override
    public ExpressionCombiner getExpressionCombiner() {
        return new ExpressionCombinerImpl();
    }

    @Override
    public RequirementBuilder newRequirementBuilder(String namespace) {
        return new RequirementBuilderImpl(namespace);
    }

**Comment by bjhargrave** [comment]
> I am trying to compile just bndlib, tag 3.1.0, stand alone. Compiling java code under bnd/biz.aQute.bndlib/src only nothing else.

It is not meant to be built this way. It is meant to be built as part of this complete git repo using the gradle build. If you want to skip building other things, then use the gradle target `:biz.aQute.bndlib:release`

```
./gradlew :biz.aQute.bndlib:release
```

This will build the necessary dependents and then build biz.aQute.bndlib placing the built bundle in `dist/bundles`.

**Comment by wltjr** [comment]
I am creating a Gentoo ebuild for bndlib, thus need to be able to build stand alone. I will look into what else is missing. Though not sure how building other parts will create the missing methods that need to be implemented. Unless its using some internal OSGI vs official. There is no support for building with Gradle on Gentoo, so that cannot be used.

**Comment by rotty3000** [comment]
But you have Java obviously, why don't you just execute the ./gradlew (just
bash + Java) which installes gradle local to the project then uses it.

- Ray

**Comment by djencks** [comment]
IIRC Genroo ebuilds require use of only gentoo provided infrastructure and can’t use the normal build systems of projects.

I’d suggest, if in line with ebuild setup requirements, running the 

./gradlew :biz.aQute.bndlib:release

command BJ suggested and looking at what it does, or at least what projects it builds in what order, and trying to build each one in ebuild, in that order, with the same dependencies as used by the gradle build.

If ideological restrictions prevent you from using gradlew as a research tool, you could ask for someone not so restricted to run the command and provide you with the output, and you might be able to guess at enough to figure out the correct order and dependencies from just that.

Out of curiosity, do ebuild policies allow you to make a gradle ebuild and then use the ebuild-supplied gradle in further ebuilds?  If so I’d guess this would be the path that will encounter the most help from the community.  I haven’t seen any java communities that have enthusiastically supported ditching their carefully customized build system to make it easier or even possible to help Gentoo.

thanks
david jencks

**Comment by wltjr** [comment]
I am going to do that, thought gradle was packaged in a binary/-bin package on gentoo and it is not. Gentoo has and is building an older version, but things have changed since.
https://github.com/gentoo/gentoo/blob/master/dev-java/bndlib/bndlib-2.1.0.ebuild

I will build bndlib using the gradle build system. But I fail to see how it will address the issue I have run into. No other sources will make those files implement the new methods that were added. When I compile using an older OSGI, 5.0.0 or 4.x it has TONS of build errors. Thus it clearly needs the newer OSGI 6.0.0, but hasn't implemented some methods in that. That will be interesting to see how building with gradle addresses that.

Thus far I have not seen anyway that gradle can be integrated into Gentoo, nor Maven. There have been many attempts to integrate Maven. The thing is portage does its own dependency resolution. It has to download and build any dependencies. Ivy is much friendlier, as it just wraps around ant. So I can bypass the ivy stuff, and keep the ant build system. But that does not work for maven or gradle sadly.

I do not expect any Java project to change their packaging at all. I am not looking to change how bnd or bndlib is built. The idea is not to help Gentoo but the other way around. Not many are building Java from source, and most just grab a pre-build binary. Many projects ship binaries as dependencies rather than building the application and all depenencies from source. There are benefits to building things from source, you discover issues, API changes, and lots of things.

Just sad that over the years Java on gentoo has fallen behind so much its not contributing in ways it could be. Last I was messing with packaging Zookeeper, i was having to update its code to newer Netty, as it still used one in a JBoss package, Netty 3.x, legacy vs current or version in development.

Gentoo can help further projects when its current, but so much Java stuff is not, sadly... Anyway let me build with gradle completely outside of portage, and see what happens there.

**Comment by wltjr** [comment]
Of course builds fine with gradle, but ships with so many versions and other *osgi*.jar files. Not sure which ones its using to build against.

**Comment by wltjr** [comment]
In https://github.com/bndtools/bnd/blob/master/biz.aQute.bndlib/bnd.bnd I see some of the deps,
-buildpath:  \
	org.osgi.service.component.annotations;version=1.3.0,\
	org.osgi.service.metatype.annotations;version=1.3.0,\
	osgi.core;version=@5,\
	osgi.cmpn;version=@5,\

But if I use osgi.cmpn version 5.0.0 I get errors about missing annotation classes that are in 6.0.0. Maybe I need to use more than once version of OSGI when building bndlib. I have not tried that yet.

This is me building with OSGI 5.0.0
http://dpaste.com/1SE5MK5

This is with OSGI 6.0.0
http://dpaste.com/023WFG9

**Comment by wltjr** [comment]
I am having a hard time finding
org.osgi.service.component.annotations;version=1.3.0,\
org.osgi.service.metatype.annotations;version=1.3.0,\

Seems those are in 6.0.0 per
http://njbartlett.name/2015/08/17/osgir6-declarative-services.html

Not sure why that calls it DS 1.3, when there seems to be no such thing. I can find it on maven, but not osgi.org.
http://mvnrepository.com/artifact/org.osgi/org.osgi.service.component.annotations/1.3.0

Seems I need annotations from 6.0.0, but OSGI core and compendium from 5.0.0

**Comment by wltjr** [comment]
Ok that seems to be the problem, I found these two jars
./cnf/repo/org.osgi.service.component.annotations/org.osgi.service.component.annotations-1.3.0.jar
./cnf/repo/org.osgi.service.metatype.annotations/org.osgi.service.metatype.annotations-1.3.0.jar

Not sure where they come from or the 1.3 version, they are clearly in 
https://osgi.org/download/r6/osgi.cmpn-6.0.0.jar

dev ~ # unzip -l /usr/share/osgi-compendium-6/lib/osgi-compendium.jar | grep org.osgi.service.component.annotations
        0  02-15-2016 17:41   org/osgi/service/component/annotations/
      406  02-15-2016 17:41   org/osgi/service/component/annotations/Activate.class
     1353  02-15-2016 17:41   org/osgi/service/component/annotations/Component.class
     1395  02-15-2016 17:41   org/osgi/service/component/annotations/ConfigurationPolicy.class
      410  02-15-2016 17:41   org/osgi/service/component/annotations/Deactivate.class
      406  02-15-2016 17:41   org/osgi/service/component/annotations/Modified.class
     1268  02-15-2016 17:41   org/osgi/service/component/annotations/FieldOption.class
     1553  02-15-2016 17:41   org/osgi/service/component/annotations/Reference.class
     1468  02-15-2016 17:41   org/osgi/service/component/annotations/ReferenceCardinality.class
     1296  02-15-2016 17:41   org/osgi/service/component/annotations/ReferencePolicy.class
     1342  02-15-2016 17:41   org/osgi/service/component/annotations/ReferencePolicyOption.class
     1384  02-15-2016 17:41   org/osgi/service/component/annotations/ReferenceScope.class
     1425  02-15-2016 17:41   org/osgi/service/component/annotations/ServiceScope.class
      248  02-15-2016 17:41   org/osgi/service/component/annotations/package-info.class
dev ~ # unzip -l /usr/share/osgi-compendium-6/lib/osgi-compendium.jar | grep org.osgi.service.metatype.annotations
        0  02-15-2016 17:41   org/osgi/service/metatype/annotations/
      939  02-15-2016 17:41   org/osgi/service/metatype/annotations/AttributeDefinition.class
     1821  02-15-2016 17:41   org/osgi/service/metatype/annotations/AttributeType.class
      544  02-15-2016 17:41   org/osgi/service/metatype/annotations/Designate.class
      409  02-15-2016 17:41   org/osgi/service/metatype/annotations/Icon.class
      735  02-15-2016 17:41   org/osgi/service/metatype/annotations/ObjectClassDefinition.class
      429  02-15-2016 17:41   org/osgi/service/metatype/annotations/Option.class
      247  02-15-2016 17:41   org/osgi/service/metatype/annotations/package-info.class


I guess I can create another package on gentoo, but the name and version will be coming out of no where, assuming like osgi-ds-1.3 or something like that. But that does not seem to exist on its own. Aside from Maven, you cannot download a osgi ds 1.3 jar anywhere else, surely not from OSGI.

Thus someone re-packaged that, and bndlib is using 6.0.0 annotations with 5.0.0. Which is why I cannot get it to build, as it does not implement all 6.0.0 stuff, but does need the annotations from that version.

**Comment by rotty3000** [comment]
OSGi specifications are semantically versioned.

This means that what you are seeing here is quite literally correct:

org.osgi.service.component.annotations;version=1.3.0,\
org.osgi.service.metatype.annotations;version=1.3.0,\
osgi.cmpn;version=@5 <https://github.com/5>,\

This is saying:

for the API org.osgi.service.component.annotations use version 1.3.0 (from
comp R6)
for the API org.osgi.service.metatype.annotations use version 1.3.0 (from
comp R6)
for any other compendium API's, use what's found in the R5 bundle.

This is possible because OSGi takes great care to ensure peer
specifications don't  break each other's compatibility and so it's
completely possible to use one compendium spec at a higher version than
other compendium specs.

You will note that as of R6, OSGi Alliance has provided individually
packaged specification bundles (called companion bundles) and that's why
you see these new bundles versioned at their spec version rather than at
their release version.

- Ray


On Tue, Feb 16, 2016 at 9:39 PM, William L. Thomson Jr. <
notifications@github.com> wrote:

> I am having a hard time finding
> org.osgi.service.component.annotations;version=1.3.0,\
> org.osgi.service.metatype.annotations;version=1.3.0,\
>
> Seems those are in 6.0.0 per
> http://njbartlett.name/2015/08/17/osgir6-declarative-services.html
>
> Not sure why that calls it DS 1.3, when there seems to be no such thing. I
> can find it on maven, but not osgi.org.
>
> http://mvnrepository.com/artifact/org.osgi/org.osgi.service.component.annotations/1.3.0
>
> Seems I need annotations from 6.0.0, but OSGI core and compendium from
> 5.0.0
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/bndtools/bnd/issues/1322#issuecomment-184983475>.
>



-- 
*Raymond Augé* <http://www.liferay.com/web/raymond.auge/profile>
 (@rotty3000)
Senior Software Architect *Liferay, Inc.* <http://www.liferay.com>
 (@Liferay)
Board Member & EEG Co-Chair, OSGi Alliance <http://osgi.org> (@OSGiAlliance)

**Comment by rotty3000** [comment]
One more thing,

The jars like osgi.cmpn.jar or osgi.enterprise.jar, etc.

Those were _never intended_ to mean "all these specs within have to be used
together at these specific versions."

Those were merely convenience "companion" bundles where all the specs of a
release were placed in a package for easy consumption. It was expected that
people would open those up and extract the parts they needed and either
consume or implement those (you can clearly see this with projects like
Eclipse Equinox and Apache Felix).

- Ray

On Tue, Feb 16, 2016 at 9:57 PM, Raymond Auge <raymond.auge@liferay.com>
wrote:

> OSGi specifications are semantically versioned.
>
> This means that what you are seeing here is quite literally correct:
>
> org.osgi.service.component.annotations;version=1.3.0,\
> org.osgi.service.metatype.annotations;version=1.3.0,\
> osgi.cmpn;version=@5 <https://github.com/5>,\
>
> This is saying:
>
> for the API org.osgi.service.component.annotations use version 1.3.0 (from
> comp R6)
> for the API org.osgi.service.metatype.annotations use version 1.3.0 (from
> comp R6)
> for any other compendium API's, use what's found in the R5 bundle.
>
> This is possible because OSGi takes great care to ensure peer
> specifications don't  break each other's compatibility and so it's
> completely possible to use one compendium spec at a higher version than
> other compendium specs.
>
> You will note that as of R6, OSGi Alliance has provided individually
> packaged specification bundles (called companion bundles) and that's why
> you see these new bundles versioned at their spec version rather than at
> their release version.
>
> - Ray
>
>
> On Tue, Feb 16, 2016 at 9:39 PM, William L. Thomson Jr. <
> notifications@github.com> wrote:
>
>> I am having a hard time finding
>> org.osgi.service.component.annotations;version=1.3.0,\
>> org.osgi.service.metatype.annotations;version=1.3.0,\
>>
>> Seems those are in 6.0.0 per
>> http://njbartlett.name/2015/08/17/osgir6-declarative-services.html
>>
>> Not sure why that calls it DS 1.3, when there seems to be no such thing.
>> I can find it on maven, but not osgi.org.
>>
>> http://mvnrepository.com/artifact/org.osgi/org.osgi.service.component.annotations/1.3.0
>>
>> Seems I need annotations from 6.0.0, but OSGI core and compendium from
>> 5.0.0
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/bndtools/bnd/issues/1322#issuecomment-184983475>.
>>
>
>
>
> --
> *Raymond Augé* <http://www.liferay.com/web/raymond.auge/profile>
>  (@rotty3000)
> Senior Software Architect *Liferay, Inc.* <http://www.liferay.com>
>  (@Liferay)
> Board Member & EEG Co-Chair, OSGi Alliance <http://osgi.org>
> (@OSGiAlliance)
>



-- 
*Raymond Augé* <http://www.liferay.com/web/raymond.auge/profile>
 (@rotty3000)
Senior Software Architect *Liferay, Inc.* <http://www.liferay.com>
 (@Liferay)
Board Member & EEG Co-Chair, OSGi Alliance <http://osgi.org> (@OSGiAlliance)

**Comment by wltjr** [comment]
Where are these jars coming from?
./cnf/repo/org.osgi.service.component.annotations/org.osgi.service.component.annotations-1.3.0.jar
./cnf/repo/org.osgi.service.metatype.annotations/org.osgi.service.metatype.annotations-1.3.0.jar

Because the contents of those are also in
./cnf/repo/osgi.cmpn/osgi.cmpn-6.0.0.jar

Literally same classes, in the bundled/companion jar, and also in stand alone.

I am not clear where the 1.3 version is coming from. How can you tell what version those annotations are in osgi.cmpn-6.0.0.jar?

It would seem really odd to create a osgi-ds package on gentoo, and have it provide classes that are also provided in osgi-compendium/osgi-cmpn.

**Comment by wltjr** [comment]
Ah seeing the packageinfo which has version in it. But there are 56 of those in the one osgi.cmpn-6.0.0.jar but some are the same, not sure if they go together or not. No clue how to package this now. I guess I need to split it up into some other packages or something... What fun...

**Comment by wltjr** [comment]
Thanks Ray for pointing me in the right direction! Turns out this is a total mess on Gentoo. Seems it was assumed the package names should be based on the jar/bundle names. Which it turns out there are duplicates in some, like the util stuff is in both compendium and enterprise. Clearly this was meant to be many packages, but it is not on Gentoo. Working on correcting that now Packaging Java applications on Gentoo is hardly trivial. None the less I got it building :)
https://github.com/Obsidian-StudiosInc/os-xtoo/blob/master/dev-java/bndlib/bndlib-3.1.0.ebuild

Thanks to all! Going to close now for real, it was an accident before.
